### PR TITLE
SAAS-3064: Added ECR Repository Encrypted Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/ecr/ecrRepositoryEncrypted.spec.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            ecrRepositoryEncrypted.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('No ECR Repository  found');
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to list ECR Repository', function (done) {
+            const cache = createCache(null, null, null, { message: "Unable to list Timestream Databases encryption" });
```
Comment (sadeed12345) [inline plugins/aws/ecr/ecrRepositoryEncrypted.spec.js:164]:
```suggestion
            const cache = createCache(null, null, null, { message: "Unable to list ECR Repository encryption" });
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                ['ecr', 'describeRepositories', region]);
+               
+            if (!describeRepositories) return rcb();
+            
+            if (describeRepositories.err || !describeRepositories.data) {
+                helpers.addResult(results, 3, `Unable to query ECR Repository: ${helpers.addError(describeRepositories)}`, region);
+                return rcb();
+            }
+            
+
```
Comment (sadeed12345) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:46]:
```suggestion
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    `Unable to list KMS keys: ${helpers.addError(listKeys)}`, region);
+                return rcb();
+            }
+
+            for (let repository of describeRepositories.data) {
+                if (!repository.repositoryArn) continue;
+
+                let resource = repository.repositoryArn;
+
+                if (repository.encryptionConfiguration.encryptionType &&
```
Comment (dipsubha06) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:65]:
```suggestion
                if ( repository.encryptionConfiguration &&
                repository.encryptionConfiguration.encryptionType &&
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                            `Unable to query KMS key: ${helpers.addError(describeKey)}`,
+                            region, kmsKey);
+                        continue;
+                    }
+
+                    currentEncryptionLevel = helpers.getEncryptionLevel(describeKey.data.KeyMetadata, helpers.ENCRYPTION_LEVELS);
+                } else {
+                    helpers.addResult(results, 2,
+                        'ECR Repository does not have encryption enabled',
+                        region, resource);
```
Comment (dipsubha06) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:87]: should not we have continue here? If this flow goes below currentEncryptionLevel value wont be set and result can be something unexpected.

**Review by dipsubha06 — CHANGES_REQUESTED** [review]

File: exports.js [hunk 5]
```diff
@@ -156,6 +156,7 @@ module.exports = {
         'efsEncryptionEnabled'          : require(__dirname + '/plugins/aws/efs/efsEncryptionEnabled.js'),
 
         'ecrRepositoryPolicy'           : require(__dirname + '/plugins/aws/ecr/ecrRepositoryPolicy.js'),
+        'ecrRepositoryEncrypted'           : require(__dirname + '/plugins/aws/ecr/ecrRepositoryEncrypted.js'),
```
Comment (AkhtarAmir) [inline exports.js:159]:
```suggestion
        'ecrRepositoryEncrypted'         : require(__dirname + '/plugins/aws/ecr/ecrRepositoryEncrypted.js'),
```

File: exports.js [hunk 6]
```diff
@@ -156,6 +156,7 @@ module.exports = {
         'efsEncryptionEnabled'          : require(__dirname + '/plugins/aws/efs/efsEncryptionEnabled.js'),
 
         'ecrRepositoryPolicy'           : require(__dirname + '/plugins/aws/ecr/ecrRepositoryPolicy.js'),
+        'ecrRepositoryEncrypted'         : require(__dirname + '/plugins/aws/ecr/ecrRepositoryEncrypted.js'),
```
Comment (AkhtarAmir) [inline exports.js:159]:
```suggestion
        'ecrRepositoryEncrypted'        : require(__dirname + '/plugins/aws/ecr/ecrRepositoryEncrypted.js'),
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'ECR Repository Encrypted',
+    category: 'ECR',
+    domain: 'Containers',
+    description: 'Ensures that the ECR Repository is Encrypted',
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:8]:
```suggestion
    description: 'Ensure that the data in ECR repository is encrypted using desired encryption level.',
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    more_info: 'Amazon ECR Repository encrypts your data with AWS-manager sse keys by default.' +
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:9]:
```suggestion
    more_info: 'Amazon ECR encrypts your images stored in the repository with AWS-managed KMS keys by default.' +
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'ECR Repository Encrypted',
+    category: 'ECR',
+    domain: 'Containers',
+    description: 'Ensure that the data in ECR repository is encrypted using desired encryption level.',
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:8]:
```suggestion
    description: 'Ensure that the images in ECR repository are encrypted using desired encryption level.',
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    description: 'Ensure that the images in ECR repository are encrypted using desired encryption level.',
+    more_info: 'Amazon ECR encrypts your images stored in the repository with AWS-managed KMS keys by default.' +
+               'Encrypt your files using customer-managed keys in order to gain more granular control over encryption/decryption process.',
+    recommended_action: 'Create ECR Repository with customer-manager keys (CMKs).',
+    link: 'https://docs.aws.amazon.com/AmazonECR/latest/userguide/Repositories.html',
+    apis: ['ECR:describeRepositories', 'KMS:describeKey', 'KMS:listKeys'],
+    settings: {
+        ecr_repository_desired_encryption_level: {
+            name: 'ECR Repository Encryption',
+            description: 'In order (lowest to highest) sse=AWS-managed AES256;awskms=AWS-managed KMS; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:17]:
```suggestion
            description: 'In order (lowest to highest) awskms=AWS-managed KMS; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+               
+            if (!describeRepositories) return rcb();
+            
+            if (describeRepositories.err || !describeRepositories.data) {
+                helpers.addResult(results, 3, `Unable to query ECR Repository: ${helpers.addError(describeRepositories)}`, region);
+                return rcb();
+            }
+            
+            if (!describeRepositories.data.length) {
+                helpers.addResult(results, 0, 'No ECR Repository  found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:47]:
```suggestion
                helpers.addResult(results, 0, 'No ECR repositories found', region);
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        var currentEncryptionLevel;
+
+        async.each(regions.ecr, function(region, rcb){
+            var describeRepositories = helpers.addSource(cache, source,
+                ['ecr', 'describeRepositories', region]);
+               
+            if (!describeRepositories) return rcb();
+            
+            if (describeRepositories.err || !describeRepositories.data) {
+                helpers.addResult(results, 3, `Unable to query ECR Repository: ${helpers.addError(describeRepositories)}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:42]:
```suggestion
                helpers.addResult(results, 3, `Unable to query ECR repositories: ${helpers.addError(describeRepositories)}`, region);
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'ECR Repository Encrypted',
+    category: 'ECR',
+    domain: 'Containers',
+    description: 'Ensure that the images in ECR repository are encrypted using desired encryption level.',
+    more_info: 'Amazon ECR encrypts your images stored in the repository with AWS-managed KMS keys by default.' +
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:9]:
```suggestion
    more_info: 'By default, Amazon ECR uses server-side encryption with Amazon S3-managed encryption keys which encrypts your data at rest using an AES-256 encryption algorithm.' +
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    'Use customer-managed keys instead, in order to gain more granular control over encryption/decryption process.',
+    recommended_action: 'Create ECR Repository with customer-manager keys (CMKs).',
+    link: 'https://docs.aws.amazon.com/AmazonECR/latest/userguide/Repositories.html',
+    apis: ['ECR:describeRepositories', 'KMS:describeKey', 'KMS:listKeys'],
+    settings: {
+        ecr_repository_desired_encryption_level: {
+            name: 'ECR Repository Encryption',
+            description: 'In order (lowest to highest) awskms=AWS-managed KMS; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
+            regex: '^(awskms|awscmk|externalcmk|cloudhsm)$',
+            default: 'awscmk'
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:19]:
```suggestion
            name: 'ECR Repository Encryption',
            description: 'In order (lowest to highest) sse=AES-256; awskms=AWS-managed KMS; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
            regex: '^(sse|awskms|awscmk|externalcmk|cloudhsm)$',
            default: 'awscmk'
```

File: plugins/aws/ecr/ecrRepositoryEncrypted.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+                    var keyId = kmsKey.split('/')[1] ? kmsKey.split('/')[1] : kmsKey;
+
+                    var describeKey = helpers.addSource(cache, source,
+                        ['kms', 'describeKey', region, keyId]);  
+
+                    if (!describeKey || describeKey.err || !describeKey.data || !describeKey.data.KeyMetadata) {
+                        helpers.addResult(results, 3,
+                            `Unable to query KMS key: ${helpers.addError(describeKey)}`,
+                            region, kmsKey);
+                        continue;
```
Comment (AkhtarAmir) [inline plugins/aws/ecr/ecrRepositoryEncrypted.js:80]:
```suggestion
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by dipsubha06 — APPROVED** [review]

**Pull request merged**